src/snowflake/snowpark/_internal/ast/ast.py

```diff
@@ -0,0 +1,130 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
+#
+
+import base64
+import itertools
+import sys
+import uuid
+from collections import namedtuple
+from dataclasses import dataclass
+from typing import Callable
+
+import snowflake.snowpark._internal.proto.generated.ast_pb2 as proto
+from snowflake.snowpark.version import VERSION
+
+# The current AST version number (generated by the DSL).
+CLIENT_AST_VERSION = proto.__Version__.MAX_VERSION
+
+
+@dataclass
+class TrackedCallable:
+    """
+    Several Snowpark APIs that deal with stored procedures and user-defined functions accept callables as arguments.
+    This class is a pair of a callable and an ID that is used to reference it in the AST. Distinct objects get distinct IDs.
+    It is undesirable for the same callable to have multiple IDs due to constraints in other parts of the system.
+    """
+
+    var_id: int
+    func: Callable
+
+
```

Comment:
nit: should this be a dataclass as well? it makes not much sense to use a namedtuple.

Why?
-> why should this be true at all? `isinstance(SerializedBatch, tuple)`?

Reply:
The original signature of `AstBatch.flush()` advertised the return type as a `Tuple[str, str]`.

If I make this a `dataclass`, all consumers of that serialized batch (not many, but a few places in the code) will need coordinated changes. I think we should do that in a follow-up PR.

Reply:
I think namedtuple should be fine since we are not going to inherit on it and it only add a name to the existing tuple.